enterprise/tools/rbesxs/rbesxs.go

```diff
@@ -0,0 +1,222 @@
+	if len(bbUrls) == 0 {
+		return ""
+	}
+	invocationIds := make([]string, len(bbUrls))
+	for i, bbUrl := range bbUrls {
+		split := strings.Split(bbUrl, "/")
+		invocationIds[i] = split[len(split)-1]
+	}
+	return "SELECT invocation_id, duration_usec, action_cache_hits, action_cache_misses, cas_cache_hits, cas_cache_uploads, total_download_size_bytes, total_upload_size_bytes FROM Invocations WHERE (invocation_id='" + strings.Join(invocationIds, "' OR invocation_id='") + "');"
+}
+
+func main() {
+	flag.Parse()
+
+	if *iterations < 0 {
+		log.Fatal("--iterations cannot be < 0")
+	}
+	if !(*environment == "dev" || *environment == "prod") {
+		log.Fatal("--environment must be 'dev' or 'prod'")
+	}
+	if *target == "" {
+		log.Fatal("--target required")
+	}
+	if *controlRepo == "" || *testRepo == "" {
+		log.Fatal("--test_repo and --control_repo must be specified")
+	}
+
+	ctx := context.Background()
+
+	control := make([]string, *iterations)
+	test := make([]string, *iterations)
+
+	failures := 0
+	controlTag := randomString()
+	testTag := randomString()
+
+	for i := 0; i < *iterations; i++ {
+		if failures > *iterations/2 {
+			log.Fatal("Failure rate is too high, aborting!")
+		}
+
+		remoteInstanceName := randomString()
+		if *doSetup {
+			bazelClean(ctx, *controlRepo)
+			setupCmd := newCommand(*target, *environment, *apiKey, remoteInstanceName, "", strings.Split(*setupFlags, ","))
+			log.Infof("iteration %d: running setup command: %s", i, strings.Join(setupCmd.Arguments, " "))
+			result := commandutil.Run(ctx, setupCmd, *controlRepo, nil, nil)
+			if result.ExitCode != 0 {
+				failures++
+				log.Infof("setup command failed, retrying: %s", string(result.Stderr))
+				i--
+				continue
+			}
+		}
+
+		bazelClean(ctx, *controlRepo)
+		controlCmd := newCommand(*target, *environment, *apiKey, remoteInstanceName, controlTag, strings.Split(*controlFlags, ","))
+		log.Infof("iteration %d: running control command: %s", i, strings.Join(controlCmd.Arguments, " "))
+		result := commandutil.Run(ctx, controlCmd, *controlRepo, nil, nil)
+		if result.ExitCode != 0 {
+			failures++
+			log.Infof("control command failed, retrying: %s", string(result.Stderr))
+			i--
+			continue
+		}
+		control[i] = getBuildbuddyUrlOrDie(string(result.Stderr))
+		log.Infof("iteration %d: control finished: %s", i, control[i])
+
+		// TODO(iain): it might be interesting/useful to support running test
+		// and control in parallel to isolate time effects.
+		remoteInstanceName = randomString()
+		if *doSetup {
+			bazelClean(ctx, *testRepo)
+			setupCmd := newCommand(*target, *environment, *apiKey, remoteInstanceName, "", strings.Split(*setupFlags, ","))
+			log.Infof("iteration %d: running setup command: %s", i, strings.Join(setupCmd.Arguments, " "))
+			result = commandutil.Run(ctx, setupCmd, *controlRepo, nil, nil)
+			if result.ExitCode != 0 {
+				failures++
+				log.Infof("setup command failed, retrying: %s", string(result.Stderr))
+				i--
+				continue
+			}
+		}
```

Comment:
Is it intentional that the setup build runs on every iteration? I would think that you could do the setup build once per repo at the very beginning (or maybe 2-3 times per repo if you really want to make sure that every executor gets warmed up). After that, the executors should be pretty warmed up and ready for the "actual" test phase

Reply:
It is intentional. Each control and test run uses a distinct remote instance names to isolate them from each other, and for the test I was running I wanted one setup build to take place with that instance name before running the test build. We could certainly make this behavior customizable in the future, but this was the behavior I wanted for my test, so it's what I added support for.